File: tools/src/data/affordability.ts
/**
 * candidate_affordability — given the units already in a list and a points
 * budget, price the cheapest next copy of each candidate unit and flag whether
 * it still fits. Powers the list-builder's "sort cheapest-first / grey out the
 * unaffordable" catalog affordance, and is exposed as a cross-impl primitive so
 * the maths is pinned by `conformance/affordability/`.
 *
 * The cost of "one more copy" is ordinal-aware: 11e prices some datasheets by
 * army ordinal (see {@link baseUnitPoints}), so the next copy of a datasheet you
 * already field twice may cost more than the first. `nextCopyCost` is the
 * cheapest *entry point* — the minimum over the unit's points tiers of
 * `baseUnitPoints(unit, tier.models, nextOrdinal)` — i.e. taking it at its
 * smallest legal size, at the ordinal it would enter the army.
 *
 * Mirror of `crates/wh40kdc/src/data/affordability.rs`,
 * `python/src/wh40kdc/data/affordability.py`, `go/affordability.go`.
 *
 * @packageDocumentation
 */
import type { BattleSize } from "../import/types.js";
import type { Faction, Unit } from "../generated.js";
import type { Dataset } from "./dataset.js";
import type { UnitView } from "./entities.js";
import { pointsLimitForBattleSize } from "./battle-sizes.js";
import { hostPointsTiers, hostUnitPoints } from "./pricing.js";

/** One unit already in the list (fixes the running total + per-datasheet ordinals). */
export interface AffordabilityUnit {
  unitId: string;
  modelCount: number;
  enhancementId?: string | null;
}

/** Compact input shared by {@link candidateAffordability} and the runner op. */
export interface AffordabilitySpec {
  factionId: string | null;
  battleSize: BattleSize | null;
  /** Explicit points limit; overrides the battle-size default when set. */
  pointsLimitOverride?: number | null;
  units: AffordabilityUnit[];
  /** Units to price; defaults to every unit in `factionId` when omitted. */
  candidateUnitIds?: string[] | null;
}

/** Affordability verdict for one candidate unit. */
export interface CandidateCost {
  unitId: string;
  nextCopyCost: number;
  affordable: boolean;
}

/** The cheapest cost to field one more copy of `view` at army ordinal `nextOrdinal`. */
function cheapestNextCopy(
  view: UnitView,
  nextOrdinal: number,
  hostFaction?: Faction | null,
): number {
  const tiers = hostPointsTiers(view.raw as Unit, hostFaction);
  if (tiers.length === 0) return 0;
  let min = Infinity;
  for (const t of tiers) {
    const cost = hostUnitPoints(view.raw, t.models, nextOrdinal, hostFaction);
    if (cost < min) min = cost;
  }
  return min === Infinity ? 0 : min;
}

/**
 * Price the cheapest next copy of each candidate and flag affordability against
 * the remaining budget. Returns one {@link CandidateCost} per candidate that
 * resolves in the dataset, sorted ascending by `(nextCopyCost, unitId)` —
 * deterministic for conformance.
 */
export function candidateAffordability(spec: AffordabilitySpec, dataset: Dataset): CandidateCost[] {
  const resolve = (unitId: string): UnitView | undefined => {
    if (!unitId) return undefined;
    if (spec.factionId) {
      const scoped = dataset.units.getInFaction(unitId, spec.factionId);
      if (scoped) return scoped;
    }
    return dataset.units.getAny(unitId);
  };

  // Running total of the current list (ordinal-aware) + enhancement costs.
  // Host-aware: foreign units with an allied_points entry for this army price
  // from that entry (see hostPointsTiers).
  const hostFaction = spec.factionId ? dataset.factions.get(spec.factionId)?.raw : undefined;
  const ordinals = new Map<string, number>();
  let spent = 0;
  for (const u of spec.units) {
    const view = resolve(u.unitId);
    if (!view) continue;
    const ord = (ordinals.get(u.unitId) ?? 0) + 1;
    ordinals.set(u.unitId, ord);
    spent += hostUnitPoints(view.raw, u.modelCount, ord, hostFaction);
    if (u.enhancementId) spent += dataset.enhancements.get(u.enhancementId)?.cost ?? 0;
  }

  const limit =
    spec.pointsLimitOverride != null ? spec.pointsLimitOverride : pointsLimitForBattleSize(spec.battleSize);
  const remaining = limit == null ? Infinity : limit - spent;

  // Candidate set: explicit list, else every unit in the faction.
  const candidateIds =
    spec.candidateUnitIds ??
    (spec.factionId ? dataset.units.byFaction(spec.factionId).map((v) => v.id) : []);

  const out: CandidateCost[] = [];
  for (const unitId of candidateIds) {
    const view = resolve(unitId);
    if (!view) continue;
    const nextOrdinal = (ordinals.get(unitId) ?? 0) + 1;
    const nextCopyCost = cheapestNextCopy(view, nextOrdinal, hostFaction);
    out.push({ unitId: view.id, nextCopyCost, affordable: nextCopyCost <= remaining });
  }
  out.sort((a, b) => (a.nextCopyCost === b.nextCopyCost ? a.unitId.localeCompare(b.unitId) : a.nextCopyCost - b.nextCopyCost));
  return out;
}

File: crates/wh40kdc/src/data/affordability.rs
//! `candidate_affordability` — given the units already in a list and a points
//! budget, price the cheapest next copy of each candidate unit and flag whether
//! it still fits. Powers the list-builder's "sort cheapest-first / grey out the
//! unaffordable" catalog affordance, and is exposed as a cross-impl primitive so
//! the maths is pinned by `conformance/affordability/`.
//!
//! The cost of "one more copy" is ordinal-aware: 11e prices some datasheets by
//! army ordinal (see [`base_unit_points`]), so the next copy of a datasheet you
//! already field twice may cost more than the first. `next_copy_cost` is the
//! cheapest *entry point* — the minimum over the unit's points tiers of
//! `base_unit_points(unit, tier.models, next_ordinal)` — i.e. taking it at its
//! smallest legal size, at the ordinal it would enter the army.
//!
//! Mirror of `tools/src/data/affordability.ts`.

use std::collections::HashMap;

use crate::data::battle_sizes::points_limit_for_battle_size;
use crate::data::pricing::{host_points_tiers, host_unit_points};
use crate::generated::{Faction, Unit};
use crate::import::BattleSize;
use crate::Dataset;

/// One unit already in the list (fixes the running total + per-datasheet ordinals).
#[derive(Debug, Clone)]
pub struct AffordabilityUnit {
    pub unit_id: String,
    pub model_count: u64,
    pub enhancement_id: Option<String>,
}

/// Compact input shared by [`candidate_affordability`] and the runner op.
#[derive(Debug, Clone)]
pub struct AffordabilitySpec {
    pub faction_id: Option<String>,
    pub battle_size: Option<BattleSize>,
    /// Explicit points limit; overrides the battle-size default when set.
    pub points_limit_override: Option<u64>,
    pub units: Vec<AffordabilityUnit>,
    /// Units to price; defaults to every unit in `faction_id` when `None`.
    pub candidate_unit_ids: Option<Vec<String>>,
}

/// Affordability verdict for one candidate unit.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CandidateCost {
    pub unit_id: String,
    pub next_copy_cost: u64,
    pub affordable: bool,
}

/// The cheapest cost to field one more copy of `unit` at army ordinal
/// `next_ordinal` — the minimum over the tiers in effect for `host_faction`'s
/// army, taking it at its smallest legal size. Zero when there are no tiers.
fn cheapest_next_copy(unit: &Unit, next_ordinal: u64, host_faction: Option<&Faction>) -> u64 {
    host_points_tiers(unit, host_faction)
        .into_iter()
        .map(|t| host_unit_points(unit, t.models, next_ordinal, host_faction))
        .min()
        .unwrap_or(0)
}

/// Price the cheapest next copy of each candidate and flag affordability against
/// the remaining budget. Returns one [`CandidateCost`] per candidate that
/// resolves in the dataset, sorted ascending by `(next_copy_cost, unit_id)` —
/// deterministic for conformance. Mirror of TS `candidateAffordability`.
pub fn candidate_affordability(spec: &AffordabilitySpec, dataset: &Dataset) -> Vec<CandidateCost> {
    let faction = spec.faction_id.as_deref();
    let resolve = |unit_id: &str| -> Option<&Unit> {
        if unit_id.is_empty() {
            return None;
        }
        faction
            .and_then(|f| {
                dataset
                    .units
                    .by_faction(f)
                    .into_iter()
                    .find(|u| u.id.as_str() == unit_id)
            })
            .or_else(|| dataset.units.get_any(unit_id))
    };

    // Running total of the current list (ordinal-aware) + enhancement costs.
    // Host-aware: foreign units with an `allied_points` entry for this army
    // price from that entry (see `host_points_tiers`).
    let host_faction = faction.and_then(|f| dataset.factions.get(f));
    let mut ordinals: HashMap<String, u64> = HashMap::new();
    let mut spent: u64 = 0;
    for u in &spec.units {
        let Some(view) = resolve(&u.unit_id) else {
            continue;
        };
        let ord = ordinals.entry(u.unit_id.clone()).or_insert(0);
        *ord += 1;
        spent += host_unit_points(view, u.model_count, *ord, host_faction);
        if let Some(enh_id) = &u.enhancement_id {
            spent += dataset
                .enhancements
                .get(enh_id)
                .map(|e| e.cost)
                .unwrap_or(0);
        }
    }

    let limit = match spec.points_limit_override {
        Some(l) => Some(l),
        None => points_limit_for_battle_size(spec.battle_size),
    };
    // `None` limit ⇒ unbounded budget (the TS `Infinity`). Compute remaining as a
    // signed value so an over-budget list yields a negative remaining (everything
    // unaffordable) rather than wrapping.
    let remaining: Option<i64> = limit.map(|l| l as i64 - spent as i64);

    // Candidate set: explicit list, else every unit in the faction.
    let candidate_ids: Vec<String> = match &spec.candidate_unit_ids {
        Some(ids) => ids.clone(),
        None => match faction {
            Some(f) => dataset
                .units
                .by_faction(f)
                .into_iter()
                .map(|u| u.id.as_str().to_string())
                .collect(),
            None => Vec::new(),
        },
    };

    let mut out: Vec<CandidateCost> = Vec::new();
    for unit_id in candidate_ids {
        let Some(view) = resolve(&unit_id) else {
            continue;
        };
        let next_ordinal = ordinals.get(&unit_id).copied().unwrap_or(0) + 1;
        let next_copy_cost = cheapest_next_copy(view, next_ordinal, host_faction);
        let affordable = match remaining {
            None => true,
            Some(r) => (next_copy_cost as i64) <= r,
        };
        out.push(CandidateCost {
            unit_id: view.id.as_str().to_string(),
            next_copy_cost,
            affordable,
        });
    }
    out.sort_by(|a, b| {
        a.next_copy_cost
            .cmp(&b.next_copy_cost)
            .then_with(|| a.unit_id.cmp(&b.unit_id))
    });
    out
}
